src/lightning/data/streaming/dataset_optimizer.py

```diff
@@ -600,18 +618,29 @@ def prepare_item(item_metadata: T) -> Any:
                 if current_total == num_items:
                     break
 
-        for w in self.workers:
-            w.join(0)
+        num_nodes = _get_num_nodes()
+
+        # TODO: Understand why it hangs.
+        if num_nodes == 1:
+            for w in self.workers:
+                w.join(0)
+
+        print("Workers are finished.")
 
         cache_dir = os.path.join(_get_cache_folder(), self.name)
+
+        chunks = [file for file in os.listdir(cache_dir) if file.endswith(".bin")]
+        if chunks and self.delete_cached_files and self.remote_dst_dir:
+            raise RuntimeError(f"All the chunks should have been deleted. Found {chunks}")
+
         merge_cache = Cache(cache_dir, chunk_bytes=1)
-        num_nodes = _get_num_nodes()
         node_rank = _get_node_rank()
-        merge_cache.merge(self.num_workers, node_rank if num_nodes > 1 else None)
+        merge_cache._merge_no_wait(node_rank if num_nodes > 1 else None)
         self._upload_index(cache_dir, num_nodes, node_rank)
-
         print("Finished data processing!")
-        print()
+
+        if num_nodes > 1:
+            os._exit(0)
```

Comment:
Why are we exiting the program? This is just the end of the run() method.

Reply:
I will try to remove it but I wanted to make sure it won't hang for any weird reasons.